### PR TITLE
Faster inference

### PR DESCRIPTION
**Pull request by claudiosv**

### CODE REVIEW

File: lmwrapper/structs.py [hunk 1]
```diff
@@ -152,6 +152,7 @@ def to_default_string_prompt(self) -> str:
 @dataclass
 class LmPrediction:
     completion_text: str
+    clean_completion_text: str
```
Comment (DNGros) [inline lmwrapper/structs.py:155]: I would say "clean" text should be the norm, not the exception. This is modeling off the openai endpoints which default to giving you clean text, and keeps in the spirit of making the OpenAI and Huggingface and etc all intermixable (in this design the user has to know whether they should choose the clean_completion_text or the completion_text).

In this design it would be a variable for something like `raw_completion_text`. 

I am a bit confused by the need for this though. What models are giving "unclean" outputs?

File: test/test_docs.py [hunk 2]
```diff
@@ -4,6 +4,9 @@
 from pathlib import Path
 import pytest
 
+from lmwrapper.secret_manage import SecretEnvVar
+SKIP_OPENAI = not SecretEnvVar("OPENAI_API_KEY").is_readable()
```
Comment (DNGros) [inline test/test_docs.py:8]: I'm not a huge fan of this. OpenAI endpoints is still central to the package. Tests should not pass if they don't work. I can see an argument for an environoment variable or config setting that lets a developer explicitly disable them, but this shouldn't be the norm.

File: lmwrapper/huggingface_wrapper.py [hunk 3]
```diff
@@ -2,23 +2,110 @@
 from functools import cached_property
 from typing import Union, List, Any
 
-from transformers.models.auto.tokenization_auto import PreTrainedTokenizerFast
-
+from torch import device
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:5]: this should be in the torch import check catcher below

File: lmwrapper/huggingface_wrapper.py [hunk 4]
```diff
@@ -105,24 +216,28 @@ def new_call(*args, **kwargs):
...
 
         self._model.forward = old_forward
 
         s = generation_output.sequences[0]
-        text = self._tokenizer.decode(s[len(encoded_input['input_ids'][0]):])
+        text = self._tokenizer.decode(
+            s[len(encoded_input["input_ids"][0]) :],
+        )
+
+        cleaned_text = self._tokenizer.decode(
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:235]: I think we should prefer to only running decode once. I think we can make the unclean version lazy (still kinda confused the situation there)

File: lmwrapper/huggingface_wrapper.py [hunk 5]
```diff
@@ -76,19 +167,39 @@ def _predict_maybe_cached(self, prompt: LmPrompt) -> Union[LmPrediction, List[Lm
...
+        not (
+            hasattr(self._model, "use_bettertransformer")
+            and self._model.use_bettertransformer is True
         )
-        #output = self._model(**encoded_input)
-        #text = self._tokenizer.decode(output[0])
+
+        encoded_input = self._tokenizer(
+            self._tokenizer.bos_token
+            + prompt.text,  # TODO: not sure why bos token is prepended for all models?
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:178]: At least in the GPT-2 case this was to ensure that we get a probability back for the first token, thus better mirroring what we get from the openai endpoints. Not sure how this interacts with the expectations of other models. We might need to think through logic here. There is an unimplemented flag in the prompt for this, but in efforts to making the package a kinda "no brains, sensable defaults" package, might need to think through this a bit more

File: test/test_huggingface.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+CUDA_UNAVAILABLE = not torch.cuda.is_available()
+SMALL_GPU = CUDA_UNAVAILABLE or torch.cuda.mem_get_info()[0] < 17179869184  # 16GB
+
+if SMALL_GPU:
+    ALL_MODELS = ["distilgpt2", "gpt2", "Salesforce/codet5p-220m", "Salesforce/codegen2-1B"]
+    CAUSAL_MODELS = ["distilgpt2", "gpt2", "Salesforce/codegen2-1B"]
+else:
+    ALL_MODELS = ["distilgpt2", "gpt2", "Salesforce/codet5p-6b", "Salesforce/codegen2-3_7B"]
+    CAUSAL_MODELS = ["distilgpt2", "gpt2", "Salesforce/codegen2-3_7B"]
```
Comment (DNGros) [inline test/test_huggingface.py:18]: I don't get the reasoning here for altering behavior based off the machine you are running on GPU's size. I wouldn't expect a 1B model to work when a 3.7B in the same model family. (unless you are maybe testing like multigpu distributed running. But it is not clear that is happening here). GPT-2 is intentionally the smallest popular causal model I could find so fast download and to test. What's the upside here of mixing in these other models?

File: lmwrapper/huggingface_wrapper.py [hunk 7]
```diff
@@ -179,9 +296,179 @@ def _gather_logprobs_from_logits(
...
+
 def get_huggingface_lm(
-    model_name: str
+    model: str,
+    runtime: Runtime = Runtime.PYTORCH,
+    precision: torch.dtype = torch.float32,
+) -> HuggingfacePredictor:
+    _kwargs = {}
+    model_class = AutoModelForCausalLM
+    if model.startswith("Salesforce/codegen"):
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:318]: is there a way to catch this without specific model names or does it fail too late or mysteriously?

File: lmwrapper/huggingface_wrapper.py [hunk 8]
```diff
@@ -179,9 +296,179 @@ def _gather_logprobs_from_logits(
...
+        if runtime == Runtime.BETTER_TRANSFORMER:
+            raise Exception(
+                "WARNING BetterTransformer breaks CodeGen models with AutoClass. Please use a different model or runtime."
+            )
+        else:
+            _kwargs = {"trust_remote_code": True, "revision": "main", "device_map": "auto"}
+    elif model.startswith("Salesforce/codet5") and not model.endswith("b"):
+        model_class = T5ForConditionalGeneration
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:326]: why isn't the autoclass functionality in huggingface not able to handle this?

File: lmwrapper/huggingface_wrapper.py [hunk 9]
```diff
@@ -179,9 +296,179 @@ def _gather_logprobs_from_logits(
...
+            model_class.from_pretrained(
+                model_name, torch_dtype=precision, **_kwargs
+            )
+        )
+        warmup_model(model, tokenizer, device=torch_device)
+
+    return HuggingfacePredictor(tokenizer, model)
+
+
+def warmup_model(
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:437]: what's going on with this method? Why is this needed?

File: lmwrapper/secret_manage.py [hunk 10]
```diff
@@ -23,7 +23,7 @@ def is_readable(self) -> bool:
 class SecretFile(SecretInterface):
     def __init__(self, path: Path):
         if not path.exists():
-            raise FileNotFoundError(f"Secret file {path} not found")
+            print(f"Secret file {path} not found")
```
Comment (DNGros) [inline lmwrapper/secret_manage.py:26]: What was the reason for this change? Probably this check either shouldn't be here at all, or it should be exception (vs cluttering the user's stdout)

File: lmwrapper/huggingface_wrapper.py [hunk 11]
```diff
@@ -76,19 +167,39 @@ def _predict_maybe_cached(self, prompt: LmPrompt) -> Union[LmPrediction, List[Lm
...
         )
-        #output = self._model(**encoded_input)
-        #text = self._tokenizer.decode(output[0])
+
+        encoded_input = self._tokenizer(
+            self._tokenizer.bos_token
+            + prompt.text,  # TODO: not sure why bos token is prepended for all models?
+            return_tensors="pt",
+            padding=True,  # TODO: not all models have a padding token
+            truncation=True,
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:181]: I don't love this. Seems like using this truncation might lead to unpredictable behaviour. There is abandoned code around lmwrapper providing utilities for prompt trimming (https://github.com/DNGros/lmwrapper/blob/master/lmwrapper/prompt_trimming.py). Need to rethink this. Generally seems like should default to failing when exceeding prompt.

File: test/test_models_common.py [hunk 12]
```diff
@@ -1,14 +1,17 @@
 from lmwrapper.huggingface_wrapper import get_huggingface_lm
 from lmwrapper.openai_wrapper import get_open_ai_lm
 import pytest
+from lmwrapper.secret_manage import SecretEnvVar
 
 from lmwrapper.structs import LmPrompt
 import math
+SKIP_OPENAI = not SecretEnvVar("OPENAI_API_KEY").is_readable()
```
Comment (DNGros) [inline test/test_models_common.py:8]: see above about not autoskipping these

File: test/test_huggingface.py (hunk 6 above)
Comment (DNGros) [reply test/test_huggingface.py:18]: (also, if going to do this, it seems like they should be seperate skipable tests so more transparent in test results whats actually being tested)

File: lmwrapper/huggingface_wrapper.py [hunk 13]
```diff
@@ -179,9 +296,179 @@ def _gather_logprobs_from_logits(
...
+    runtime: Runtime = Runtime.PYTORCH,
+    precision: torch.dtype = torch.float32,
+    _kwargs: dict = {},
 ) -> HuggingfacePredictor:
-    tokenizer = AutoTokenizer.from_pretrained(model_name)
-    model = AutoModelForCausalLM.from_pretrained(model_name)
-    return HuggingfacePredictor(tokenizer, model)
\ No newline at end of file
+    torch_device = get_accelerator()
+    if runtime == Runtime.PYTORCH:
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:352]: can we make these branches seperate functions to avoid such a long method here?

File: test/test_huggingface.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+from lmwrapper.huggingface_wrapper import Runtime, get_huggingface_lm
+from lmwrapper.structs import LmPrompt
+import torch
+
+
+CUDA_UNAVAILABLE = not torch.cuda.is_available()
+SMALL_GPU = CUDA_UNAVAILABLE or torch.cuda.mem_get_info()[0] < 17179869184  # 16GB
+
+if SMALL_GPU:
+    ALL_MODELS = ["distilgpt2", "gpt2", "Salesforce/codet5p-220m", "Salesforce/codegen2-1B"]
```
Comment (DNGros) [inline test/test_huggingface.py:14]: I'm surprised these t5 style models work. It is not clear they are tested. Also might need to think through this some more as they Seq2Seq models might expect to be different decoder only models and thus have different wrapper endpoint (or be more explicit on how we are making them act like a autoregressive model, kinda like how we are able how the wrapper lets autoregressive models act like a dialog model)

**Review by DNGros — COMMENTED** [review]
Cool, a lot of good stuff! Made an initial pass at a review. Can try to take another look at this later as this PR is doing a lot of stuff.

File: test/test_docs.py (hunk 2 above)
Comment (DNGros) [reply test/test_docs.py:8]: * them = the openai tests

File: lmwrapper/structs.py (hunk 1 above)
Comment (claudiosv) [reply lmwrapper/structs.py:155]: It looks like you were cleaning up tokenization spaces in `remove_special_chars_from_tokens`. I agree the default should be output without special tokens.

File: test/test_docs.py (hunk 2 above)
Comment (claudiosv) [reply test/test_docs.py:8]: I agree, but the secret manager does not expose a suitable surface, instead the logic to load the secret from env or file is in `get_open_ai_lm` method. Ideally CI can still run tests if the secret is provided.

File: lmwrapper/huggingface_wrapper.py (hunk 3 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:5]: Thanks! Missed that

File: lmwrapper/huggingface_wrapper.py (hunk 4 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:235]: Agreed

File: test/test_huggingface.py (hunk 6 above)
Comment (claudiosv) [reply test/test_huggingface.py:18]: It seems stupid but it has a very good reason. Look at the configs for [codet5p-16b](https://huggingface.co/Salesforce/codet5p-16b/blob/main/config.json) and [codet5p-220m](https://huggingface.co/Salesforce/codet5p-220m/blob/main/config.json). They actually are different models. The larger one runs custom code from `CodeT5pEncoderDecoderModel` and the smaller one just uses `T5ForConditionalGeneration`. I want to be able to run the tests locally for small models and run the big models on a machine with large GPU(s). There is an argument to be made for running the small models on the big machine too, but these tests already take forever to run.

File: lmwrapper/huggingface_wrapper.py (hunk 7 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:318]: I think it fails too late / mysteriously. One idea to download the model config and handle it, but the required/supported args are specific to certain models. AutoModel fails in some cases for ORT and BetterTransformers if not handled earlier.

File: lmwrapper/huggingface_wrapper.py (hunk 8 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:326]: Apparently not, see example code [here](https://huggingface.co/Salesforce/codet5p-220m#how-to-use). The error you get with AutoModel is ```ValueError: Unrecognized configuration class <class 'transformers.models.t5.configuration_t5.T5Config'> for this kind of AutoModel: AutoModelForCausalLM.```

File: lmwrapper/huggingface_wrapper.py (hunk 9 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:437]: This isn't really needed for non-ORT models. For ORT, this creates the fused kernel(s). My understanding is the input size will determine the subsequent max input sizes possible (so that will also need some work).

File: lmwrapper/secret_manage.py (hunk 10 above)
Comment (claudiosv) [reply lmwrapper/secret_manage.py:26]: This check shouldn't be here imo, that's what the `is_readable` method should be used for. Without an openai key you just get crashing tests.

File: lmwrapper/huggingface_wrapper.py (hunk 11 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:181]: Makes sense.

File: lmwrapper/huggingface_wrapper.py (hunk 13 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:352]: Agreed.

File: test/test_models_common.py (hunk 12 above)
Comment (claudiosv) [reply test/test_models_common.py:8]: Maybe an env variable like "SKIP_OPENAI" makes sense.

File: test/test_huggingface.py (hunk 14 above)
Comment (claudiosv) [reply test/test_huggingface.py:14]: Agreed.

File: lmwrapper/huggingface_wrapper.py [hunk 15]
```diff
@@ -313,20 +315,28 @@ def get_huggingface_lm(
     runtime: Runtime = Runtime.PYTORCH,
     precision: torch.dtype = torch.float32,
 ) -> HuggingfacePredictor:
-    _kwargs = {}
+    _kwargs = {"trust_remote_code": True}
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:318]: this feels like a dubious default. If huggingface/whoever thinks it should not be a default, not sure we want to expose users of our wrapper to these risks (the audience here is people who don't actually want to deal with all the intricacies of HF and just want reasonable defaults). This might need to be as a user specified flag and ideally caught with an informative error message.

File: lmwrapper/huggingface_wrapper.py (hunk 7 above)
Comment (DNGros) [reply lmwrapper/huggingface_wrapper.py:320]: If auto model fails we can catch that and explain common reasons for the failure. Another idea might be to do a small call at the end (similar to the warmup) and catch a error that indicates this there? Idk, I know for our purposes in Synthegrator these are the key models, but this style of checks feel like a game where constantly have to keep hard coding names. I can be ok with this for now if just want to get something in to be able to run experiments, but it is omethign to fix later.

File: test/test_huggingface.py (hunk 6 above)
Comment (DNGros) [reply test/test_huggingface.py:18]: hm... ok, weird. I'd say test_huggingface file can try to use a minimal set of smallest possible models then a seperate path for "huggingface_zoo_tests" where we just try a mix of models we know we want to support. These can go another a separate mark that is not run by default (like the SO answer https://stackoverflow.com/a/61193490 . It can be either a `slow` flag, or like a `run_zoo` flag). This is not set normally or in the CI. What do you think?

File: lmwrapper/huggingface_wrapper.py [hunk 16]
```diff
@@ -208,24 +209,24 @@ def _predict_maybe_cached(
         #   require calling the model again https://discuss.huggingface.co/t/announcement-generation-get-probabilities-for-generated-output/30075/17
         # Instead we are going to patch the model forward to log calls
 
-        old_forward = self._model.forward
-        cached_logits = []
+        # old_forward = self._model.forward
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:212]: This was a pain to get working but was needed. I suspect logprobs might just not work with the fast runtimes (at least not like this. Might work via a call at the end to run the whole sequence through one last time to get the values outside of generate)

File: lmwrapper/huggingface_wrapper.py (hunk 15 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:318]: Agreed. I will work on adding AutoConfig loading, and checking the config whether it specifies remote classes. If a remote class is specified, I think lmwrapper should raise an exception if `trust_remote_code` is False (as it should be). The problem here is that many new models have custom model classes on HF Hub, and the default behavior leads to unexpected results as the correct model class may not be loaded.

**Comment by claudiosv** [comment]
As for the warmup thing, it is only needed for TensorRT. Therefore, I have removed it from the others. I originally left it in because it was a handy debugging tool but now that I have written tests I have a way to catch failed models. See documentation regarding [warmup](https://huggingface.co/docs/optimum/onnxruntime/usage_guides/gpu#tensorrt-engine-build-and-warmup). My intention is to make the warmup function run one sequence of 1 token and then one sequence with model's max_length number of tokens. This follows the documentation ensuring that the model provided is ready for use. 

```
[...] first build the TensorRT engine with an input of small shape, and then with an input of large shape to have an engine valid for all shapes inbetween. This allows to avoid rebuilding the engine for new small and large shapes, which is unwanted once the model is deployed for inference.
```

If we don't warmup, the unexpected result to users will be that they pass in different sequences, say lengths 10,20,30 and get abysmal performance as each sequence is slightly longer than the previous, resulting in a full TensorRT model build which is very, very slow.

**Comment by claudiosv** [comment]
I wanted to add some goals and motivations for this PR as well.

1. I would like to run large models on Mariana/Houston machine for bigcalibrate experiments. To this end, I would like to support ORT CUDA and/or ORT TensorRT backends, and some models may even require 8/4bit quant.
2. We told SRI we would try CodeGen2 and CodeT5+ models. This also means that any model specific hacks such as prepending a token without the user knowing, should be avoided.
3. I would like a model instantiated by lmwrapper to be ready for use in bigcalibrate with relegating as few model specific flags/kwargs/etc to bigcalibrate. Ditto for TensorRT warmup if I can get that to work.

File: lmwrapper/structs.py (hunk 1 above)
Comment (DNGros) [reply lmwrapper/structs.py:155]: So the `remove_special_chars_from_tokens` is at the token level, not at the text level. I think the `completion_text` should just always be clean.

File: test/test_docs.py (hunk 2 above)
Comment (DNGros) [reply test/test_docs.py:8]: ok, this seems reasonable. Would you mind adding a little dummy test along the lines of
```python
def test_ci_has_secret():
     in_github_action = os.getenv("GITHUB_ACTIONS") == "true"
     if in_github_action:
         assert not SKIP_OPENAI
```

File: lmwrapper/secret_manage.py (hunk 10 above)
Comment (DNGros) [reply lmwrapper/secret_manage.py:26]: ok, agreed. Let's remove it.

File: test/test_models_common.py (hunk 12 above)
Comment (DNGros) [reply test/test_models_common.py:8]: I said above about just keeping the skip with an assert on github actions. However, I actually think a mark for if we should run the OPEN_AI ones make more sense. That also allows us eventually make a separate workflow that runs on external PRs without the secret in the environoment which will still at least run some of the tests (but can be a task for later)

File: lmwrapper/huggingface_wrapper.py (hunk 13 above)
Comment (DNGros) [reply lmwrapper/huggingface_wrapper.py:402]: (sorry, this a bit of a nit. It is fine as-is for now)

**Comment by DNGros** [comment]
> As for the warmup thing, it is only needed for TensorRT. Therefore, I have removed it from the others. I originally left it in because it was a handy debugging tool but now that I have written tests I have a way to catch failed models. See documentation regarding [warmup](https://huggingface.co/docs/optimum/onnxruntime/usage_guides/gpu#tensorrt-engine-build-and-warmup). My intention is to make the warmup function run one sequence of 1 token and then one sequence with model's max_length number of tokens. This follows the documentation ensuring that the model provided is ready for use.
> 
> > [...] first build the TensorRT engine with an input of small shape, and then with an input of large shape to have an engine valid for all shapes inbetween. This allows to avoid rebuilding the engine for new small and large shapes, which is unwanted once the model is deployed for inference.
> 
> If we don't warmup, the unexpected result to users will be that they pass in different sequences, say lengths 10,20,30 and get abysmal performance as each sequence is slightly longer than the previous, resulting in a full TensorRT model build which is very, very slow.

Got it, makes sense. Thanks for explaining. Still suggest we use the existing codepath in `HuggingfacePredictor` for the warmup as we discussed above.

**Comment by DNGros** [comment]
> I wanted to add some goals and motivations for this PR as well.
> 
>     1. I would like to run large models on Mariana/Houston machine for bigcalibrate experiments. To this end, I would like to support ORT CUDA and/or ORT TensorRT backends, and some models may even require 8/4bit quant.
> 
>     2. We told SRI we would try CodeGen2 and CodeT5+ models. This also means that any model specific hacks such as prepending a token without the user knowing, should be avoided.
> 
>     3. I would like a model instantiated by lmwrapper to be ready for use in bigcalibrate with relegating as few model specific flags/kwargs/etc to bigcalibrate. Ditto for TensorRT warmup if I can get that to work.

2. makes sense. I would still say let's get the causal ones working first before thinking about the T5.
3. got it. Seems reasonable. Agreed warmup good idea now. Also agree that idea behind the wrapper is to give sensible defaults (also to avoid kwargs whenever possible, preferring object oriented and statically checkable parameters instead)

File: lmwrapper/huggingface_wrapper.py (hunk 16 above)
Comment (DNGros) [reply lmwrapper/huggingface_wrapper.py:215]: If by "that guy" you mean the guy who catches the bugs, please do! :)

That is weird. I'll look into it a bit.

To add more context, the reason for this hack is because the compute_transition_scores method won't expose the scores of the prompt. Plus it is unclear how the varying logits wrappers affect it, when really just want to calculate the probability of the text under the model.

Comment (DNGros) [reply lmwrapper/huggingface_wrapper.py:215]: Which codegen2 model is this on?

At least on main branch I get reasonableish non-zero probabilities
```python
if __name__ == "__main__":
    lm = get_huggingface_lm('Salesforce/codegen2-1B')
    pred = lm.predict(LmPrompt("Hello World", max_tokens=1, temperature=0))
    print("pred", pred)
    print("completion_text", pred.completion_text)
    print("full_text", pred.get_full_text())
    print("completion_logprobs", np.exp(pred.completion_logprobs))
    print("---- transition scores ----")
    transition_scores = lm._model.compute_transition_scores(
        pred.metad.sequences, pred.metad.scores, normalize_logits=True,
    )
    print(pred.metad.scores)

    input_length = len(pred.prompt_tokens) + 1
    generated_tokens = pred.metad.sequences[:, input_length:]
    for tok, score in zip(generated_tokens[0], transition_scores[0]):
        # | token | token string | logits | probability
        print(
            f"| {tok:5d} | {lm._tokenizer.decode(tok):8s} | {score.numpy():.4f} | {np.exp(score.numpy()):.2%}")
```
```text
pred HuggingfacePrediction(completion_text='o', prompt=LmPrompt(text='Hello World', max_tokens=1, stop=None, logprobs=1, temperature=0.0, top_p=0.9, presence_penalty=0.0, num_completions=1, cache=None, echo=False, add_bos_token=True), metad=GreedySearchDecoderOnlyOutput(sequences=tensor([[50256, 15496,  2159,    78]]), scores=(tensor([[-4.9619, -4.2172, -0.0112,  ..., -1.2814, -3.5225, -1.0767]]),), attentions=None, hidden_states=None), _prompt_encoding={'input_ids': tensor([[50256, 15496,  2159]]), 'attention_mask': tensor([[1, 1, 1]])}, _tokens=['Hello', 'ĠWorld', 'o'], _log_probs=tensor([-17.3223, -11.7011,  -2.3423]))
completion_text o
full_text Hello Worldo
completion_logprobs tensor([0.0961])
---- transition scores ----
(tensor([[-4.9619, -4.2172, -0.0112,  ..., -1.2814, -3.5225, -1.0767]]),)
|    78 | o        | -2.3422 | 9.61%
```
(though the chosen token is a bit wacky)
